### ISSUE TITLE
Add TypeScript support

### ISSUE THREAD

**Issue by ebeloded** [material-components/material-components-web-react#48]
This issue will inevitably come up at some point: the community will need type definitions to reliably utilize the components in TypeScript projects.

Since this project is still at an early stage - maybe instead of creating and maintaining a separate package with definitions, it would make more sense to transition the project to TypeScript? This will provide a guarantee that the code, including changes and new features, are always in sync.

**Comment by patrick91** [comment]
Yeah, TypeScript by default will really be helpful, it will be much easier the maintain, especially since most of the components won't be super complex anyway

**Comment by bonniezhou** [comment]
Thanks for the feedback! It's not currently on our roadmap but we hear your feedback. We will add it to our tracker.

**Comment by kraenhansen** [comment]
I would love to help this transition happen.

**Comment by pspeter3** [comment]
I'm also excited and happy for this to happen.

**Comment by nickmccurdy** [comment]
Would type contributions be accepted for existing components, or should we wait?

**Comment by pspeter3** [comment]
I think that's a valid concern. One way we can mitigate that is by generating the TypeScript definitions automatically from the `PropTypes` and `defaultProps`. For example, Button has:

```js
Button.propTypes = {
  raised: PropTypes.bool,
  unelevated: PropTypes.bool,
  outlined: PropTypes.bool,
  disabled: PropTypes.bool,
  unbounded: PropTypes.bool,
  initRipple: PropTypes.func,
  className: PropTypes.string,
  icon: PropTypes.element,
  children: PropTypes.string,
};

Button.defaultProps = {
  raised: false,
  unelevated: false,
  outlined: false,
  disabled: false,
  unbounded: false,
  initRipple: () => {},
  className: '',
  icon: null,
  children: '',
};
```

The TypeScript would be

```ts
export interface Props {
    raised?: boolean;
    unelevated?: boolean;
    outline?: boolean;
    disabled?: boolean;
    initRipple?: Function;
    className?: string;
    icon?: React.ReactElement<any>
    children?: string;
}
```

Would you be open to a script which added type definitions?

**Comment by nickmccurdy** [comment]
Maybe we can base types off those from [RMWC](https://github.com/jamesmfriedman/rmwc), which are autogenerated from Flow type declarations based on MDC.

**Comment by jamesmfriedman** [comment]
Yeah for RMWC I actually wrote the code using Flow (more forgiving than TS) and wrote a ridiculously simple conversion to Typescript that happens during build time that relies heavily on convention.

But the bonus is first class Typings in the two most popular type checkers for react.

**Comment by lynnjepsen** [comment]
@pspeter3 that sounds reasonable....i guess the `export interface Props` file something we would include in NPM? I also guess we would generate this Typescript file each time we deploy to NPM? So that if we change our Button.jsx source code such that there are new propTypes, we also generate a new TypeScript file for this Props?

I don't think I would want types based of RMWC. Theres no guarantee that each components API matches 1:1 at any point in time. Ideally they are as close as possible, but close as possible isn't good enough to pass a compiler.

**Comment by pspeter3** [comment]
We would generate a DTS per package most likely. I think it depends on how developers are expected to import components. I would be happy to drive this work but would appreciate guidance on how you would want the script to integrate with the code base.

**Comment by jamesmfriedman** [comment]
@lynnjepsen you don't need to publish tsx files, just generate the .d.ts typescript definition files. I'm sure there is an AST parser you can use that reads proptypes, dump it to json, and then in turn use those to generate the definition.

I went the other direction, generated the tsx files and then used the tsc compiler to generate the definitions for me.

**Comment by pspeter3** [comment]
The script plan was to actually just look at the `PropType` object on the components and then generate a DTS file. No need to have an AST parser as far as I know.

**Comment by jamesmfriedman** [comment]
Sure, but you're talking about runtime code vs some compilation or build step that needs to introspect it. PropTypes at runtime don't give you any information to get back to what you input. Last I checked, they produce anonymous functions that get called for checking values which have know indicator of what type they are actually checking.

Anyhow I'm sure you'll figure it out. I've been down the PropTypes path before, there are a few packages out there to help you extract information from them, and they do so by monkey patching the prop types lib.

**Comment by moog16** [comment]
- [ ] Button
- [ ] Card
- [ ] Checkbox
- [ ] Chips
- [ ] Drawer
- [ ] Fab
- [ ] Floating Label
- [ ] Layout Grid
- [ ] Line Ripple
- [ ] List
- [ ] Material Icon
- [ ] Menu Surface
- [ ] Notched Outline
- [ ] Ripple
- [ ] Select
- [ ] Switch
- [ ] Tab
- [ ] Tab Bar
- [ ] Tab Indicator
- [ ] Text Field
- [ ] Top App Bar
